<a href="https://colab.research.google.com/github/Shopping-Yuan/ML2021HW/blob/Shopping_vscode_branch/HW4/HW04_modified.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [5]:
import shutil
shutil.copyfile('/content/drive/MyDrive/Dataset.zip','/content/Dataset.zip')

import zipfile
with zipfile.ZipFile('/content/Dataset.zip','r') as zip_ref:
  zip_ref.extractall('/content/')

In [15]:
def parse_args():
  """arguments"""
  config = {
    "data_dir": "/content/Dataset",
    "save_path": "model.ckpt",
    "batch_size": 32,
    "n_workers": 0,
    "valid_steps": 2000,
    "warmup_steps": 1000,
    "save_steps": 10000,
    "total_steps": 100000,
  }

  return config

#Dataset

In [7]:
import os
import json
import torch
import random
from pathlib import Path
from torch.utils.data import Dataset
from torch.nn.utils.rnn import pad_sequence


class myDataset(Dataset):
  def __init__(self, data_dir, segment_len=128):
    self.data_dir = data_dir
    self.segment_len = segment_len

    # Load the mapping from speaker neme to their corresponding id.
    # Path package can use / to join path
    #part 1-1
    mapping_path = Path(data_dir) / "mapping.json"
    mapping = json.load(mapping_path.open())
    self.speaker2id = mapping["speaker2id"]
    #self.speaker2id = {'id10001': 0, 'id10005': 1, ...... 'id11247': 598, 'id11251': 599}

    #part 1-2
    # Load metadata of training data.
    metadata_path = Path(data_dir) / "metadata.json"
    metadata = json.load(open(metadata_path))["speakers"]

    # Get the total number of speaker.
    self.speaker_num = len(metadata.keys())
    self.data = []
    for speaker in metadata.keys():
    #['id10473', 'id10328', 'id10382',......'id10864']
      for utterances in metadata[speaker]:
      # for example : metadata[speaker] = 'id10473'
      # utterances = {'feature_path': 'uttr-5c88b2f1803449789c36f14fb4d3c1eb.pt', 'mel_len': 652}......
        self.data.append([utterances["feature_path"], self.speaker2id[speaker]])
        # for example : ['uttr-5c88b2f1803449789c36f14fb4d3c1eb.pt',217]

  def __len__(self):
    return len(self.data)

  def __getitem__(self, index):
    feat_path, speaker = self.data[index]
    #[utterances["feature_path"], self.speaker2id[speaker]

    # Load preprocessed mel-spectrogram.
    # https://pytorch.org/docs/stable/generated/torch.load.html
    # using torch.load to open pt file . By default , tensors are loaded to GPU if they are GPU tensor.
    # part 2
    mel = torch.load(os.path.join(self.data_dir, feat_path))

    # Segmemt mel-spectrogram into "segment_len" frames.
    if len(mel) > self.segment_len:
      # Randomly get the starting point of the segment.
      start = random.randint(0, len(mel) - self.segment_len)
      # Get a segment with "segment_len" frames.
      mel = torch.FloatTensor(mel[start:start+self.segment_len])
    else:
      mel = torch.FloatTensor(mel)
    # Turn the speaker id into long for computing loss later.
    speaker = torch.FloatTensor([speaker]).long()
    return mel, speaker

  def get_speaker_number(self):
    return self.speaker_num

In [ ]:
#data_dir = parse_args()["data_dir"]
#part 1-1
#mapping.json checcking
#mapping_path = Path(data_dir) / "mapping.json"
#mapping = json.load(mapping_path.open())
#speaker2id = mapping["speaker2id"]
#print(type(mapping)) #<class 'dict'>
#print(mapping.keys()) #dict_keys(['speaker2id', 'id2speaker'])
#print(mapping.values()) #dict_values([{'id10001': 0, 'id10005': 1, ...... 'id11247': 598, 'id11251': 599},
             #{"0": "id10001", ......'599':  'id11251'}])
#print(mapping['speaker2id']['id10473']) #217

#part 1-2
#metadata.json checcking
#metadata_path = Path(data_dir) / "metadata.json"
#metadata = json.load(open(metadata_path))["speakers"]
#print(type(metadata)) #<class 'dict'>
#print(len(metadata.keys()),metadata.keys()) #600 dict_keys(['id10473', 'id10328', 'id10382',......'id10864'])
#print(len(metadata.values()),len(list(metadata.values())[0]),list(metadata.values())[0])
#600 176 [{'feature_path': 'uttr-5c88b2f1803449789c36f14fb4d3c1eb.pt', 'mel_len': 652},......
#{'feature_path': 'uttr-caf385f0032140e0add0bbde9a6595cd.pt', 'mel_len': 414}]
#print(len(metadata.values()),len(list(metadata.values())[1])) #600 101

#part 2
#mel = torch.load(os.path.join(data_dir, 'uttr-5c88b2f1803449789c36f14fb4d3c1eb.pt'))
#print(type(mel)) #<class 'torch.Tensor'>
#print(mel.size()) #torch.Size([652, 40])
#print(mel[0])
# tensor([
# 2.0060,  1.9015,  1.3473,  1.5711,  1.5569,  2.3200,  3.0064,  3.3848,
# 0.5376,  0.8397,  1.5710, -0.5662, -2.1262, -1.1271, -1.4272, -3.6233,
# -1.0765, -1.8453, -3.2415, -2.3427, -1.5888, -0.4692, -0.4291, -0.4722,
# -1.8450, -1.0638, -3.2177, -3.4006, -3.2176, -3.3409, -3.1142, -4.0395,
# -3.6502, -4.2407, -4.9656, -5.1238, -5.0620, -5.4308, -6.0757, -6.6186])

In [8]:
import torch
from torch.utils.data import DataLoader, random_split
from torch.nn.utils.rnn import pad_sequence
#part 1 check collate_batch
def collate_batch(batch):
  # Process features within a batch.
  """Collate a batch of data."""
  #part 1-1 , 1-2
  mel, speaker = zip(*batch)
  # Because we train the model batch by batch, we need to pad the features in the same batch to make their lengths the same.
  mel = pad_sequence(mel, batch_first=True, padding_value=-20)    # pad log 10^(-20) which is very small value.
  # mel: (batch size, length, 40)
  #part 1-3
  return mel, torch.FloatTensor(speaker).long()


def get_dataloader(data_dir, batch_size, n_workers):
  """Generate dataloader"""
  dataset = myDataset(data_dir)
  speaker_num = dataset.get_speaker_number()
  # Split dataset into training dataset and validation dataset
  trainlen = int(0.9 * len(dataset))
  lengths = [trainlen, len(dataset) - trainlen]
  trainset, validset = random_split(dataset, lengths)

  train_loader = DataLoader(
    trainset,
    batch_size=batch_size,
    shuffle=True,
    drop_last=True,
    num_workers=n_workers,
    pin_memory=True,
    collate_fn=collate_batch,
  )
  valid_loader = DataLoader(
    validset,
    batch_size=batch_size,
    num_workers=n_workers,
    drop_last=True,
    pin_memory=True,
    collate_fn=collate_batch,
  )

  return train_loader, valid_loader, speaker_num

In [ ]:
# x = [["a",1],["b",2],["c",3]]
# print(list(zip(*x)))
# print(list(zip(("a",1),("b",2),("c",3))))

In [ ]:
#import numpy as np
#check collate_batch
#part 1-1
# * collect "sequence"function input to a tuple :
#def f(*paras):
#  print(paras)
#f(1,2,3) #(1, 2, 3)
# or collect a tuple to a "sequence"
#print(*(1,2,3)) #1 2 3

#part 1-2
#test_data = np.random.rand(2,3,4)
#test_label = np.random.rand(2,1)
#samples = (test_data,test_label)
#print(samples)
#x , y = zip((1,2),(3,4),(5,6),(7,8))
#print(x,y)

#part 1-3
#print(torch.FloatTensor([[1.],[1.]]))#tensor([[1.],[1.]])
#print(torch.FloatTensor([[1.],[1.]]).long())#tensor([[1],[1]]) implies the numbers are integer now
#print(torch.IntTensor([[1.],[1.]]))#tensor([[1],[1]], dtype=torch.int32)

In [9]:
!pip install torchinfo

In [ ]:
# x = [True,True,False,True,False,False,False,True,False,True,True,False]

# true_index_list = []

# false_index_list = []

# for i,b in enumerate(x):
#   if b == True:
#     true_index_list += [i]
#   else :
#     false_index_list += [i]
# print(true_index_list,false_index_list)

# new_list = []
# t_max = len(true_index_list)
# f_max = len(false_index_list)
# true_index = 0
# false_index = 0
# t_or_f = 0
# while (t_max * f_max)>0:
#   if t_or_f % 2 ==0:
#     new_list += true_index_list[true_index:true_index+3]
#     true_index += 3
#     t_or_f += 1
#     t_max -= 3
#   else :
#     new_list += false_index_list[false_index:false_index+3]
#     false_index += 3
#     t_or_f += 1
#     f_max -= 3
# if t_max > 0:
#   new_list += true_index_list[true_index:]
# elif f_max > 0:
#   new_list += false_index_list[false_index:]

# print(new_list)


In [16]:
import torch
import torch.nn as nn
import torch.nn.functional as F
import torchvision
from torchinfo import summary

class Classifier(nn.Module):
  def __init__(self, d_model=128, n_spks=600, dropout=0.1):
    super().__init__()
    # Project the dimension of features from that of input into d_model.
    self.prenet = nn.Linear(40, d_model)
    # TODO:
    #   Change Transformer to Conformer.
    #   https://arxiv.org/abs/2005.08100
    # embed_dim must be divisible by num_heads
    self.encoder_layer = nn.TransformerEncoderLayer(
      d_model=d_model, dim_feedforward=256, nhead=1
    )
    # batch_first = True
    # norm_first = True
    self.encoder = nn.TransformerEncoder(self.encoder_layer, num_layers=4)

    # Project the the dimension of features from d_model into speaker nums.
    self.pred_layer = nn.Sequential(
      nn.Linear(d_model, d_model),
      nn.ReLU(),
      nn.Linear(d_model, n_spks),
    )

  def forward(self, mels):
    """
    args:
      mels: (batch size, length, 40)
    return:
      out: (batch size, n_spks)
    """
    # out: (batch size, length, d_model)
    out = self.prenet(mels)
    # out: (length, batch size, d_model)
    out = out.permute(1, 0, 2)
    # The encoder layer expect features in the shape of (length, batch size, d_model).
    #out = self.encoder_layer(out)
    out = self.encoder(out)
    # out: (batch size, length, d_model)
    out = out.transpose(0, 1)
    # mean pooling
    stats = out.mean(dim=1)

    # out: (batch, n_spks)
    out = self.pred_layer(stats)
    return out


In [11]:
model = Classifier()
print(summary(model,(32,128,40)))
#print(model.state_dict().keys())
#print(model.state_dict()["encoder_layer.self_attn.in_proj_weight"].size())#torch.Size([240, 80]) (80+80+80)*80
#print(model.state_dict()["encoder_layer.self_attn.out_proj.weight"].size())#torch.Size([80, 80])

/usr/local/lib/python3.10/dist-packages/torch/nn/modules/transformer.py:282: UserWarning: enable_nested_tensor is True, but self.use_nested_tensor is False because encoder_layer.self_attn.batch_first was not True(use batch_first for better inference performance)
  warnings.warn(f"enable_nested_tensor is True, but self.use_nested_tensor is False because {why_not_sparsity_fast_path}")


Layer (type:depth-idx)                        Output Shape              Param #
Classifier                                    [32, 600]                 132,480
├─Linear: 1-1                                 [32, 128, 128]            5,248
├─TransformerEncoder: 1-2                     [128, 32, 128]            --
│    └─ModuleList: 2-1                        --                        --
│    │    └─TransformerEncoderLayer: 3-1      [128, 32, 128]            132,480
│    │    └─TransformerEncoderLayer: 3-2      [128, 32, 128]            132,480
├─Sequential: 1-3                             [32, 600]                 --
│    └─Linear: 2-2                            [32, 128]                 16,512
│    └─ReLU: 2-3                              [32, 128]                 --
│    └─Linear: 2-4                            [32, 600]                 77,400
Total params: 496,600
Trainable params: 496,600
Non-trainable params: 0
Total mult-adds (M): 20.18
Input size (MB): 0.66
Forward/backward pass s

original
parameter number :
1.MultiheadAttention : =
d_model+1(bias) * d_query +
d_model+1(bias) * d_key  +
d_model+1(bias) * d_v   +
d_v + 1(bias) * output_dim( = dmodel)
= (80+1)80 + (80+1)80 + (80+1)80 + (80+1)80 = 25920
2.layer_norm : =input + input = 80*80 = 160

first change
parameter number :
1.MultiheadAttention : =
d_model+1(bias) * d_query +
d_model+1(bias) * d_key  +
d_model+1(bias) * d_v   +
d_v + 1(bias) * output_dim( = dmodel)
= (128+1)128(3+1)= 16512

second change
parameter number :
1.MultiheadAttention : =
d_model+1(bias) * d_query +
d_model+1(bias) * d_key  +
d_model+1(bias) * d_v   +
d_v + 1(bias) * output_dim( = dmodel)
= (128+1)128(3+1)= 16512
two layers : 33024

2.layer_norm : =input + input = 80*80 = 160

3rd change
parameter number :
1.MultiheadAttention : =
d_model+1(bias) * d_query +
d_model+1(bias) * d_key  +
d_model+1(bias) * d_v   +
d_v + 1(bias) * output_dim( = dmodel)
= (128+1)128(3+1)= 16512
four layers : 66048

2.layer_norm : =input + input = 80*80 = 160


In [12]:
import math

import torch
from torch.optim import Optimizer
from torch.optim.lr_scheduler import LambdaLR


def get_cosine_schedule_with_warmup(
  optimizer: Optimizer,
  num_warmup_steps: int,
  num_training_steps: int,
  num_cycles: float = 0.5,
  last_epoch: int = -1,
):
  """
  Create a schedule with a learning rate that decreases following the values of the cosine function between the
  initial lr set in the optimizer to 0, after a warmup period during which it increases linearly between 0 and the
  initial lr set in the optimizer.

  Args:
    optimizer (:class:`~torch.optim.Optimizer`):
      The optimizer for which to schedule the learning rate.
    num_warmup_steps (:obj:`int`):
      The number of steps for the warmup phase.
    num_training_steps (:obj:`int`):
      The total number of training steps.
    num_cycles (:obj:`float`, `optional`, defaults to 0.5):
      The number of waves in the cosine schedule (the defaults is to just decrease from the max value to 0
      following a half-cosine).
    last_epoch (:obj:`int`, `optional`, defaults to -1):
      The index of the last epoch when resuming training.

  Return:
    :obj:`torch.optim.lr_scheduler.LambdaLR` with the appropriate schedule.
  """

  def lr_lambda(current_step):
    # Warmup
    if current_step < num_warmup_steps:
      return float(current_step) / float(max(1, num_warmup_steps))
    # decadence
    # cosine angle
    progress = float(current_step - num_warmup_steps) / float(
      max(1, num_training_steps - num_warmup_steps)
    )
    return max(
      0.0, 0.5 * (1.0 + math.cos(math.pi * float(num_cycles) * 2.0 * progress))
    )

  return LambdaLR(optimizer, lr_lambda, last_epoch)

In [13]:
from torch.optim import AdamW
optimizer = AdamW(model.parameters(), lr=1e-3)

scheduler = get_cosine_schedule_with_warmup(optimizer, 1000, 70000)

#print(type(scheduler.optimizer.param_groups)) #<class 'list'>
#print(len(scheduler.optimizer.param_groups))  #1
#print(type(scheduler.optimizer.param_groups[0])) #<class 'dict'>
#print(scheduler.optimizer.param_groups[0].keys())
#['params', 'lr', 'betas', 'eps', 'weight_decay', 'amsgrad', 'foreach',
#'maximize', 'capturable', 'differentiable', 'fused', 'initial_lr']
#print(scheduler.optimizer.param_groups[0]['params'])
#[Parameter containing: tensor([[......],requires_grad=True)]
#print(scheduler.optimizer.param_groups[0]['lr'])# 0.0
#print(scheduler.optimizer.param_groups[0]['betas'])# (0.9, 0.999)
#print(scheduler.optimizer.param_groups[0]['eps']) # 1e-08
#print(scheduler.optimizer.param_groups[0]['weight_decay']) #0.01
#print(scheduler.optimizer.param_groups[0]['amsgrad']) # False
#print(scheduler.optimizer.param_groups[0]['foreach']) # None

In [17]:
import torch


def model_fn(batch, model, criterion, device):
  """Forward a batch through the model."""

  mels, labels = batch
  mels = mels.to(device)
  labels = labels.to(device)

  outs = model(mels)

  loss = criterion(outs, labels)

  # Get the speaker id with highest probability.
  preds = outs.argmax(1)
  # Compute accuracy.
  accuracy = torch.mean((preds == labels).float())

  return loss, accuracy

In [18]:
from tqdm import tqdm
import torch


def valid(dataloader, model, criterion, device):
  """Validate on validation set."""

  model.eval()
  running_loss = 0.0
  running_accuracy = 0.0
  # Manually update the progress bar : step 1
  # total length
  pbar = tqdm(total=len(dataloader.dataset), ncols=0, desc="Valid", unit=" uttr")
  for i, batch in enumerate(dataloader):
    with torch.no_grad():
      loss, accuracy = model_fn(batch, model, criterion, device)
      running_loss += loss.item()
      running_accuracy += accuracy.item()
    # Manually update the progress bar : step 2
    # update each batch
    pbar.update(dataloader.batch_size)
    pbar.set_postfix(
      loss=f"{running_loss / (i+1):.2f}",
      accuracy=f"{running_accuracy / (i+1):.2f}",
    )
  # Manually update the progress bar : step 3
  pbar.close()
  model.train()

  return running_accuracy / len(dataloader)


In [19]:
from tqdm import tqdm

import torch
import torch.nn as nn
from torch.optim import AdamW
from torch.utils.data import DataLoader, random_split


# def parse_args():
#   """arguments"""
#   config = {
#     "data_dir": "/content/Dataset",
#     "save_path": "model.ckpt",
#     "batch_size": 32,
#     "n_workers": 8,
#     "valid_steps": 2000,
#     "warmup_steps": 1000,
#     "save_steps": 10000,
#     "total_steps": 70000,
#   }

#   return config


def main(
  data_dir,
  save_path,
  batch_size,
  n_workers,
  valid_steps,
  warmup_steps,
  total_steps,
  save_steps,
):
  """Main function."""
  device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
  print(f"[Info]: Use {device} now!")

  train_loader, valid_loader, speaker_num = get_dataloader(data_dir, batch_size, n_workers)
  train_iterator = iter(train_loader)
  print(f"[Info]: Finish loading data!",flush = True)

  model = Classifier(n_spks=speaker_num).to(device)
  criterion = nn.CrossEntropyLoss()
  optimizer = AdamW(model.parameters(), lr=1e-3)
  scheduler = get_cosine_schedule_with_warmup(optimizer, warmup_steps, total_steps)
  print(f"[Info]: Finish creating model!",flush = True)

  best_accuracy = -1.0
  best_state_dict = None
  # pbar has total 2000 step
  pbar = tqdm(total=valid_steps, ncols=0, desc="Train", unit=" step")
  # print train loss and accuracy every 1000 step

  loss_acc_list = []

  for step in range(total_steps):
    # Get data
    try:
      batch = next(train_iterator)
    except StopIteration:
      train_iterator = iter(train_loader)
      batch = next(train_iterator)

    loss, accuracy = model_fn(batch, model, criterion, device)
    batch_loss = loss.item()
    batch_accuracy = accuracy.item()

    # Updata model
    loss.backward()
    optimizer.step()
    scheduler.step()
    optimizer.zero_grad()
    # print train loss acc
    loss_acc_list.append((loss, accuracy))
    print_loss , print_acc = list(zip(*loss_acc_list))
    if step % 1000 == 0:
      print(f"average loss = {sum(print_loss)/len(print_loss):.4f} \
      average accuracy = {sum(print_acc)/len(print_acc):.4f})")
      loss_acc_list = []

    # Log
    # update each step (as total)
    pbar.update()
    pbar.set_postfix(
      loss=f"{batch_loss:.2f}",
      accuracy=f"{batch_accuracy:.2f}",
      step=step + 1,
    )

    # Do validation
    # valid each 2000 step
    if (step + 1) % valid_steps == 0:
      #close above pbar
      pbar.close()

      valid_accuracy = valid(valid_loader, model, criterion, device)

      # keep the best model
      if valid_accuracy > best_accuracy:
        best_accuracy = valid_accuracy
        best_state_dict = model.state_dict()
      # new pbar has total 2000 step
      pbar = tqdm(total=valid_steps, ncols=0, desc="Train", unit=" step")

    # Save the best model so far.
    if (step + 1) % save_steps == 0 and best_state_dict is not None:
      torch.save(best_state_dict, save_path)
      pbar.write(f"Step {step + 1}, best model saved. (accuracy={best_accuracy:.4f})")

  pbar.close()


if __name__ == "__main__":
  main(**parse_args())

[Info]: Use cuda now!
[Info]: Finish loading data!
[Info]: Finish creating model!


Train:   0% 2/2000 [00:00<05:55,  5.62 step/s, accuracy=0.00, loss=6.38, step=2]

average loss = 6.4123       average accuracy = 0.0000)


Train:  50% 1002/2000 [01:53<02:10,  7.62 step/s, accuracy=0.12, loss=4.58, step=1002]

average loss = 5.4925       average accuracy = 0.0497)


Train: 100% 2000/2000 [03:36<00:00,  9.24 step/s, accuracy=0.16, loss=3.58, step=2000]
Valid: 100% 6944/6944 [00:20<00:00, 343.97 uttr/s, accuracy=0.23, loss=3.71]
Train:   0% 2/2000 [00:00<04:28,  7.45 step/s, accuracy=0.31, loss=3.38, step=2002]

average loss = 4.1386       average accuracy = 0.1616)


Train:  50% 1001/2000 [01:28<01:59,  8.35 step/s, accuracy=0.19, loss=3.50, step=3001]

average loss = 3.4990       average accuracy = 0.2539)


Train: 100% 2000/2000 [03:08<00:00, 10.59 step/s, accuracy=0.28, loss=3.28, step=4000]
Valid: 100% 6944/6944 [00:19<00:00, 355.04 uttr/s, accuracy=0.35, loss=2.97]
Train:   0% 2/2000 [00:00<04:34,  7.27 step/s, accuracy=0.28, loss=3.09, step=4002]

average loss = 3.1247       average accuracy = 0.3165)


Train:  50% 1001/2000 [01:29<01:40,  9.94 step/s, accuracy=0.53, loss=2.33, step=5001]

average loss = 2.8411       average accuracy = 0.3628)


Train: 100% 2000/2000 [03:09<00:00, 10.55 step/s, accuracy=0.38, loss=2.75, step=6000]
Valid: 100% 6944/6944 [00:20<00:00, 347.10 uttr/s, accuracy=0.40, loss=2.67]
Train:   0% 2/2000 [00:00<04:55,  6.75 step/s, accuracy=0.31, loss=3.47, step=6002]

average loss = 2.6696       average accuracy = 0.3998)


Train:  50% 1001/2000 [01:32<01:52,  8.90 step/s, accuracy=0.44, loss=2.17, step=7001]

average loss = 2.4773       average accuracy = 0.4300)


Train: 100% 2000/2000 [03:11<00:00, 10.44 step/s, accuracy=0.47, loss=2.21, step=8000]
Valid: 100% 6944/6944 [00:20<00:00, 340.42 uttr/s, accuracy=0.46, loss=2.38]
Train:   0% 1/2000 [00:00<04:13,  7.90 step/s, accuracy=0.44, loss=1.82, step=8001]

average loss = 2.3524       average accuracy = 0.4570)


Train:  50% 1001/2000 [01:31<01:46,  9.39 step/s, accuracy=0.50, loss=2.30, step=9001]

average loss = 2.2392       average accuracy = 0.4735)


Train: 100% 2000/2000 [03:06<00:00, 10.71 step/s, accuracy=0.50, loss=1.49, step=1e+4]
Valid: 100% 6944/6944 [00:19<00:00, 348.09 uttr/s, accuracy=0.50, loss=2.19]
Train:   0% 1/2000 [00:00<05:24,  6.15 step/s, accuracy=0.47, loss=2.26, step=1e+4]

Step 10000, best model saved. (accuracy=0.5007)
average loss = 2.1548       average accuracy = 0.4934)


Train:  50% 1001/2000 [01:33<01:45,  9.43 step/s, accuracy=0.50, loss=2.26, step=11001]

average loss = 2.0393       average accuracy = 0.5170)


Train: 100% 2000/2000 [03:06<00:00, 10.73 step/s, accuracy=0.50, loss=1.59, step=12000]
Valid: 100% 6944/6944 [00:19<00:00, 347.95 uttr/s, accuracy=0.52, loss=2.11]
Train:   0% 1/2000 [00:00<04:01,  8.29 step/s, accuracy=0.62, loss=1.81, step=12001]

average loss = 1.9646       average accuracy = 0.5334)


Train:  50% 1001/2000 [01:33<02:03,  8.08 step/s, accuracy=0.59, loss=2.42, step=13001]

average loss = 1.8975       average accuracy = 0.5487)


Train: 100% 2000/2000 [03:04<00:00, 10.84 step/s, accuracy=0.56, loss=2.19, step=14000]
Valid: 100% 6944/6944 [00:19<00:00, 348.27 uttr/s, accuracy=0.55, loss=1.96]
Train:   0% 1/2000 [00:00<03:36,  9.25 step/s, accuracy=0.59, loss=1.35, step=14001]

average loss = 1.8195       average accuracy = 0.5653)


Train:  50% 1001/2000 [01:36<01:46,  9.40 step/s, accuracy=0.56, loss=1.97, step=15001]

average loss = 1.7583       average accuracy = 0.5790)


Train: 100% 2000/2000 [03:07<00:00, 10.67 step/s, accuracy=0.62, loss=1.45, step=16000]
Valid: 100% 6944/6944 [00:20<00:00, 339.89 uttr/s, accuracy=0.57, loss=1.87]
Train:   0% 2/2000 [00:00<04:53,  6.80 step/s, accuracy=0.59, loss=1.54, step=16002]

average loss = 1.7113       average accuracy = 0.5903)


Train:  50% 1001/2000 [01:39<01:56,  8.58 step/s, accuracy=0.50, loss=1.70, step=17001]

average loss = 1.6799       average accuracy = 0.5927)


Train: 100% 2000/2000 [03:09<00:00, 10.54 step/s, accuracy=0.56, loss=1.78, step=18000]
Valid: 100% 6944/6944 [00:19<00:00, 348.18 uttr/s, accuracy=0.58, loss=1.81]
Train:   0% 2/2000 [00:00<05:18,  6.28 step/s, accuracy=0.62, loss=1.78, step=18002]

average loss = 1.6198       average accuracy = 0.6089)


Train:  50% 1001/2000 [01:36<01:53,  8.82 step/s, accuracy=0.78, loss=1.34, step=19001]

average loss = 1.5749       average accuracy = 0.6133)


Train: 100% 2000/2000 [03:02<00:00, 10.94 step/s, accuracy=0.81, loss=1.08, step=2e+4]
Valid: 100% 6944/6944 [00:20<00:00, 345.92 uttr/s, accuracy=0.59, loss=1.73]
Train:   0% 1/2000 [00:00<05:56,  5.61 step/s, accuracy=0.78, loss=1.10, step=2e+4]

Step 20000, best model saved. (accuracy=0.5938)
average loss = 1.5281       average accuracy = 0.6271)


Train:  50% 1001/2000 [01:38<01:55,  8.67 step/s, accuracy=0.50, loss=1.82, step=21001]

average loss = 1.5170       average accuracy = 0.6284)


Train: 100% 2000/2000 [03:05<00:00, 10.79 step/s, accuracy=0.72, loss=1.21, step=22000]
Valid: 100% 6944/6944 [00:20<00:00, 335.90 uttr/s, accuracy=0.60, loss=1.75]
Train:   0% 1/2000 [00:00<03:44,  8.90 step/s, accuracy=0.62, loss=1.28, step=22001]

average loss = 1.4522       average accuracy = 0.6418)


Train:  50% 1001/2000 [01:43<01:44,  9.53 step/s, accuracy=0.53, loss=1.87, step=23001]

average loss = 1.4389       average accuracy = 0.6437)


Train: 100% 2000/2000 [03:09<00:00, 10.56 step/s, accuracy=0.53, loss=1.65, step=24000]
Valid: 100% 6944/6944 [00:20<00:00, 337.40 uttr/s, accuracy=0.63, loss=1.62]
Train:   0% 2/2000 [00:00<05:07,  6.49 step/s, accuracy=0.62, loss=1.30, step=24002]

average loss = 1.3823       average accuracy = 0.6567)


Train:  50% 1002/2000 [01:40<02:17,  7.25 step/s, accuracy=0.59, loss=1.69, step=25002]

average loss = 1.3740       average accuracy = 0.6603)


Train: 100% 2000/2000 [03:05<00:00, 10.79 step/s, accuracy=0.75, loss=1.26, step=26000]
Valid: 100% 6944/6944 [00:20<00:00, 334.90 uttr/s, accuracy=0.65, loss=1.52]
Train:   0% 1/2000 [00:00<03:14, 10.25 step/s, accuracy=0.72, loss=1.06, step=26001]

average loss = 1.2970       average accuracy = 0.6764)


Train:  50% 1001/2000 [01:41<01:50,  9.07 step/s, accuracy=0.56, loss=1.52, step=27001]

average loss = 1.3219       average accuracy = 0.6723)


Train: 100% 2000/2000 [03:07<00:00, 10.67 step/s, accuracy=0.66, loss=1.29, step=28000]
Valid: 100% 6944/6944 [00:21<00:00, 325.01 uttr/s, accuracy=0.64, loss=1.53]
Train:   0% 1/2000 [00:00<04:12,  7.90 step/s, accuracy=0.66, loss=1.13, step=28001]

average loss = 1.2351       average accuracy = 0.6883)


Train:  50% 1001/2000 [01:47<01:54,  8.70 step/s, accuracy=0.69, loss=1.19, step=29001]

average loss = 1.2683       average accuracy = 0.6857)


Train: 100% 2000/2000 [03:13<00:00, 10.33 step/s, accuracy=0.69, loss=1.26, step=3e+4]
Valid: 100% 6944/6944 [00:20<00:00, 331.01 uttr/s, accuracy=0.66, loss=1.44]
Train:   0% 1/2000 [00:00<06:03,  5.49 step/s, accuracy=0.72, loss=0.99, step=3e+4]

Step 30000, best model saved. (accuracy=0.6633)
average loss = 1.1999       average accuracy = 0.7013)


Train:  50% 1002/2000 [01:48<02:20,  7.09 step/s, accuracy=0.81, loss=0.74, step=31002]

average loss = 1.2174       average accuracy = 0.6923)


Train: 100% 2000/2000 [03:11<00:00, 10.43 step/s, accuracy=0.81, loss=0.98, step=32000]
Valid: 100% 6944/6944 [00:21<00:00, 318.83 uttr/s, accuracy=0.67, loss=1.46]
Train:   0% 1/2000 [00:00<03:46,  8.83 step/s, accuracy=0.62, loss=1.58, step=32001]

average loss = 1.1562       average accuracy = 0.7074)


Train:  50% 1001/2000 [01:48<01:57,  8.49 step/s, accuracy=0.62, loss=1.20, step=33001]

average loss = 1.1391       average accuracy = 0.7142)


Train: 100% 2000/2000 [03:12<00:00, 10.38 step/s, accuracy=0.75, loss=1.48, step=34000]
Valid: 100% 6944/6944 [00:21<00:00, 328.35 uttr/s, accuracy=0.67, loss=1.39]
Train:   0% 1/2000 [00:00<04:19,  7.70 step/s, accuracy=0.72, loss=1.13, step=34001]

average loss = 1.0910       average accuracy = 0.7232)


Train:  50% 1001/2000 [01:46<01:56,  8.56 step/s, accuracy=0.69, loss=1.22, step=35001]

average loss = 1.1151       average accuracy = 0.7181)


Train: 100% 2000/2000 [03:07<00:00, 10.65 step/s, accuracy=0.72, loss=0.87, step=36000]
Valid: 100% 6944/6944 [00:21<00:00, 325.01 uttr/s, accuracy=0.69, loss=1.34]
Train:   0% 2/2000 [00:00<05:38,  5.91 step/s, accuracy=0.75, loss=1.05, step=36002]

average loss = 1.0372       average accuracy = 0.7378)


Train:  50% 1001/2000 [01:46<01:49,  9.10 step/s, accuracy=0.75, loss=0.98, step=37001]

average loss = 1.0640       average accuracy = 0.7289)


Train: 100% 2000/2000 [03:08<00:00, 10.59 step/s, accuracy=0.66, loss=1.28, step=38000]
Valid: 100% 6944/6944 [00:21<00:00, 324.30 uttr/s, accuracy=0.69, loss=1.35]
Train:   0% 1/2000 [00:00<04:28,  7.45 step/s, accuracy=0.91, loss=0.59, step=38001]

average loss = 1.0038       average accuracy = 0.7437)


Train:  50% 1001/2000 [01:47<01:54,  8.76 step/s, accuracy=0.81, loss=0.77, step=39001]

average loss = 1.0381       average accuracy = 0.7365)


Train: 100% 2000/2000 [03:11<00:00, 10.45 step/s, accuracy=0.78, loss=0.75, step=4e+4]
Valid: 100% 6944/6944 [00:21<00:00, 317.01 uttr/s, accuracy=0.70, loss=1.32]
Train:   0% 1/2000 [00:00<06:43,  4.95 step/s, accuracy=0.69, loss=1.10, step=4e+4]

Step 40000, best model saved. (accuracy=0.7005)
average loss = 0.9639       average accuracy = 0.7513)


Train:  50% 1002/2000 [01:49<01:22, 12.13 step/s, accuracy=0.75, loss=0.79, step=41002]

average loss = 0.9929       average accuracy = 0.7468)


Train: 100% 2000/2000 [03:13<00:00, 10.35 step/s, accuracy=0.72, loss=1.27, step=42000]
Valid: 100% 6944/6944 [00:22<00:00, 312.87 uttr/s, accuracy=0.71, loss=1.25]
Train:   0% 1/2000 [00:00<04:09,  8.00 step/s, accuracy=0.72, loss=1.02, step=42001]

average loss = 0.9332       average accuracy = 0.7600)


Train:  50% 1002/2000 [01:50<01:25, 11.73 step/s, accuracy=0.91, loss=0.52, step=43002]

average loss = 0.9487       average accuracy = 0.7573)


Train: 100% 2000/2000 [03:15<00:00, 10.24 step/s, accuracy=0.75, loss=0.82, step=44000]
Valid: 100% 6944/6944 [00:22<00:00, 312.02 uttr/s, accuracy=0.71, loss=1.21]
Train:   0% 2/2000 [00:00<05:51,  5.68 step/s, accuracy=0.75, loss=1.07, step=44002]

average loss = 0.8907       average accuracy = 0.7697)


Train:  50% 1002/2000 [01:46<01:21, 12.27 step/s, accuracy=0.81, loss=0.88, step=45002]

average loss = 0.9006       average accuracy = 0.7694)


Train: 100% 2000/2000 [03:12<00:00, 10.37 step/s, accuracy=0.75, loss=1.02, step=46000]
Valid: 100% 6944/6944 [00:22<00:00, 307.21 uttr/s, accuracy=0.71, loss=1.23]
Train:   0% 1/2000 [00:00<04:04,  8.19 step/s, accuracy=0.94, loss=0.40, step=46001]

average loss = 0.8351       average accuracy = 0.7856)


Train:  50% 1001/2000 [01:44<01:18, 12.68 step/s, accuracy=0.81, loss=0.91, step=47001]

average loss = 0.8699       average accuracy = 0.7741)


Train: 100% 2000/2000 [03:11<00:00, 10.43 step/s, accuracy=0.75, loss=0.94, step=48000]
Valid: 100% 6944/6944 [00:22<00:00, 309.02 uttr/s, accuracy=0.74, loss=1.13]
Train:   0% 1/2000 [00:00<04:40,  7.13 step/s, accuracy=0.94, loss=0.28, step=48001]

average loss = 0.8163       average accuracy = 0.7877)


Train:  50% 1002/2000 [01:44<01:12, 13.73 step/s, accuracy=0.75, loss=0.87, step=49002]

average loss = 0.8107       average accuracy = 0.7918)


Train: 100% 2000/2000 [03:12<00:00, 10.39 step/s, accuracy=0.78, loss=0.94, step=5e+4]
Valid: 100% 6944/6944 [00:21<00:00, 322.23 uttr/s, accuracy=0.73, loss=1.14]
Train:   0% 1/2000 [00:00<07:30,  4.44 step/s, accuracy=0.81, loss=0.79, step=5e+4]

Step 50000, best model saved. (accuracy=0.7404)
average loss = 0.7898       average accuracy = 0.7951)


Train:  50% 1002/2000 [01:45<01:18, 12.69 step/s, accuracy=0.62, loss=1.28, step=51002]

average loss = 0.7780       average accuracy = 0.7974)


Train: 100% 2000/2000 [03:17<00:00, 10.14 step/s, accuracy=0.81, loss=0.78, step=52000]
Valid: 100% 6944/6944 [00:23<00:00, 296.30 uttr/s, accuracy=0.74, loss=1.10]
Train:   0% 1/2000 [00:00<04:41,  7.11 step/s, accuracy=0.91, loss=0.39, step=52001]

average loss = 0.7528       average accuracy = 0.8018)


Train:  50% 1002/2000 [01:41<01:32, 10.78 step/s, accuracy=0.91, loss=0.45, step=53002]

average loss = 0.7591       average accuracy = 0.7992)


Train: 100% 2000/2000 [03:13<00:00, 10.33 step/s, accuracy=0.84, loss=0.44, step=54000]
Valid: 100% 6944/6944 [00:22<00:00, 313.20 uttr/s, accuracy=0.75, loss=1.09]
Train:   0% 1/2000 [00:00<05:25,  6.15 step/s, accuracy=0.81, loss=1.17, step=54001]

average loss = 0.7224       average accuracy = 0.8109)


Train:  50% 1001/2000 [01:39<01:22, 12.10 step/s, accuracy=0.84, loss=0.53, step=55001]

average loss = 0.7072       average accuracy = 0.8138)


Train: 100% 2000/2000 [03:12<00:00, 10.37 step/s, accuracy=0.97, loss=0.23, step=56000]
Valid: 100% 6944/6944 [00:22<00:00, 311.06 uttr/s, accuracy=0.76, loss=1.05]
Train:   0% 1/2000 [00:00<03:53,  8.55 step/s, accuracy=0.88, loss=0.40, step=56001]

average loss = 0.6863       average accuracy = 0.8192)


Train:  50% 1001/2000 [01:38<01:17, 12.89 step/s, accuracy=0.94, loss=0.19, step=57001]

average loss = 0.6682       average accuracy = 0.8260)


Train: 100% 2000/2000 [03:13<00:00, 10.33 step/s, accuracy=0.88, loss=0.74, step=58000]
Valid: 100% 6944/6944 [00:22<00:00, 314.11 uttr/s, accuracy=0.76, loss=1.03]
Train:   0% 1/2000 [00:00<04:27,  7.46 step/s, accuracy=0.88, loss=0.46, step=58001]

average loss = 0.6607       average accuracy = 0.8254)


Train:  50% 1001/2000 [01:37<01:28, 11.28 step/s, accuracy=0.81, loss=0.76, step=59001]

average loss = 0.6376       average accuracy = 0.8331)


Train: 100% 2000/2000 [03:13<00:00, 10.36 step/s, accuracy=0.88, loss=0.68, step=6e+4]
Valid: 100% 6944/6944 [00:21<00:00, 320.21 uttr/s, accuracy=0.77, loss=1.02]
Train:   0% 1/2000 [00:00<07:10,  4.65 step/s, accuracy=0.84, loss=0.53, step=6e+4]

Step 60000, best model saved. (accuracy=0.7717)
average loss = 0.6269       average accuracy = 0.8340)


Train:  50% 1001/2000 [01:33<01:21, 12.32 step/s, accuracy=0.97, loss=0.19, step=61001]

average loss = 0.6007       average accuracy = 0.8404)


Train: 100% 2000/2000 [03:12<00:00, 10.40 step/s, accuracy=0.88, loss=0.42, step=62000]
Valid: 100% 6944/6944 [00:22<00:00, 314.83 uttr/s, accuracy=0.77, loss=1.00]
Train:   0% 1/2000 [00:00<04:47,  6.96 step/s, accuracy=0.81, loss=0.76, step=62001]

average loss = 0.5967       average accuracy = 0.8408)


Train:  50% 1002/2000 [01:37<01:41,  9.81 step/s, accuracy=0.84, loss=0.62, step=63002]

average loss = 0.5733       average accuracy = 0.8478)


Train: 100% 2000/2000 [03:20<00:00,  9.99 step/s, accuracy=0.84, loss=0.71, step=64000]
Valid: 100% 6944/6944 [00:21<00:00, 323.89 uttr/s, accuracy=0.77, loss=0.97]
Train:   0% 1/2000 [00:00<05:18,  6.28 step/s, accuracy=0.81, loss=0.55, step=64001]

average loss = 0.5753       average accuracy = 0.8468)


Train:  50% 1002/2000 [01:34<01:37, 10.28 step/s, accuracy=0.81, loss=0.79, step=65002]

average loss = 0.5574       average accuracy = 0.8501)


Train: 100% 2000/2000 [03:12<00:00, 10.39 step/s, accuracy=0.94, loss=0.30, step=66000]
Valid: 100% 6944/6944 [00:20<00:00, 337.82 uttr/s, accuracy=0.77, loss=0.96]
Train:   0% 1/2000 [00:00<04:16,  7.79 step/s, accuracy=0.84, loss=0.58, step=66001]

average loss = 0.5426       average accuracy = 0.8569)


Train:  50% 1001/2000 [01:33<01:37, 10.26 step/s, accuracy=0.97, loss=0.13, step=67001]

average loss = 0.5269       average accuracy = 0.8596)


Train: 100% 2000/2000 [03:17<00:00, 10.14 step/s, accuracy=0.81, loss=0.83, step=68000]
Valid: 100% 6944/6944 [00:20<00:00, 338.22 uttr/s, accuracy=0.78, loss=0.96]
Train:   0% 1/2000 [00:00<05:04,  6.57 step/s, accuracy=0.78, loss=0.63, step=68001]

average loss = 0.5148       average accuracy = 0.8616)


Train:  50% 1002/2000 [01:33<01:42,  9.76 step/s, accuracy=0.75, loss=0.74, step=69002]

average loss = 0.4904       average accuracy = 0.8693)


Train: 100% 2000/2000 [03:16<00:00, 10.19 step/s, accuracy=0.72, loss=0.79, step=7e+4]
Valid: 100% 6944/6944 [00:20<00:00, 335.90 uttr/s, accuracy=0.79, loss=0.91]
Train:   0% 1/2000 [00:00<06:45,  4.93 step/s, accuracy=0.94, loss=0.46, step=7e+4]

Step 70000, best model saved. (accuracy=0.7873)
average loss = 0.4918       average accuracy = 0.8687)


Train:  50% 1002/2000 [01:31<01:44,  9.55 step/s, accuracy=0.81, loss=0.68, step=71002]

average loss = 0.4710       average accuracy = 0.8768)


Train: 100% 2000/2000 [03:18<00:00, 10.06 step/s, accuracy=0.78, loss=0.76, step=72000]
Valid: 100% 6944/6944 [00:21<00:00, 320.54 uttr/s, accuracy=0.80, loss=0.89]
Train:   0% 1/2000 [00:00<04:55,  6.77 step/s, accuracy=0.81, loss=0.69, step=72001]

average loss = 0.4691       average accuracy = 0.8749)


Train:  50% 1001/2000 [01:36<01:39,  9.99 step/s, accuracy=0.78, loss=0.78, step=73001]

average loss = 0.4548       average accuracy = 0.8793)


Train: 100% 2000/2000 [03:26<00:00,  9.69 step/s, accuracy=0.84, loss=0.64, step=74000]
Valid: 100% 6944/6944 [00:21<00:00, 322.36 uttr/s, accuracy=0.80, loss=0.90]
Train:   0% 2/2000 [00:00<05:43,  5.82 step/s, accuracy=0.91, loss=0.37, step=74001]

average loss = 0.4576       average accuracy = 0.8793)


Train:  50% 1001/2000 [01:33<01:39, 10.00 step/s, accuracy=0.84, loss=0.42, step=75001]

average loss = 0.4333       average accuracy = 0.8861)


Train: 100% 2000/2000 [03:25<00:00,  9.75 step/s, accuracy=0.88, loss=0.30, step=76000]
Valid: 100% 6944/6944 [00:21<00:00, 327.23 uttr/s, accuracy=0.81, loss=0.85]
Train:   0% 1/2000 [00:00<05:17,  6.29 step/s, accuracy=0.97, loss=0.12, step=76001]

average loss = 0.4294       average accuracy = 0.8879)


Train:  50% 1001/2000 [01:33<01:37, 10.25 step/s, accuracy=0.97, loss=0.17, step=77001]

average loss = 0.4131       average accuracy = 0.8908)


Train: 100% 2000/2000 [03:22<00:00,  9.88 step/s, accuracy=0.91, loss=0.28, step=78000]
Valid: 100% 6944/6944 [00:20<00:00, 344.42 uttr/s, accuracy=0.81, loss=0.84]
Train:   0% 1/2000 [00:00<04:21,  7.65 step/s, accuracy=0.88, loss=0.43, step=78001]

average loss = 0.4135       average accuracy = 0.8896)


Train:  50% 1002/2000 [01:30<02:01,  8.18 step/s, accuracy=0.94, loss=0.25, step=79001]

average loss = 0.3904       average accuracy = 0.8964)


Train: 100% 2000/2000 [03:19<00:00, 10.02 step/s, accuracy=0.81, loss=0.61, step=8e+4]
Valid: 100% 6944/6944 [00:21<00:00, 328.94 uttr/s, accuracy=0.81, loss=0.85]
Train:   0% 1/2000 [00:00<07:59,  4.17 step/s, accuracy=0.84, loss=0.34, step=8e+4]

Step 80000, best model saved. (accuracy=0.8142)
average loss = 0.3940       average accuracy = 0.8962)


Train:  50% 1001/2000 [01:32<01:53,  8.76 step/s, accuracy=0.94, loss=0.29, step=81001]

average loss = 0.3731       average accuracy = 0.8986)


Train: 100% 2000/2000 [03:19<00:00, 10.02 step/s, accuracy=0.84, loss=0.57, step=82000]
Valid: 100% 6944/6944 [00:21<00:00, 328.62 uttr/s, accuracy=0.81, loss=0.85]
Train:   0% 1/2000 [00:00<03:13, 10.31 step/s, accuracy=0.94, loss=0.23, step=82001]

average loss = 0.3773       average accuracy = 0.9014)


Train:  50% 1001/2000 [01:33<01:59,  8.33 step/s, accuracy=0.97, loss=0.19, step=83001]

average loss = 0.3657       average accuracy = 0.9038)


Train: 100% 2000/2000 [03:22<00:00,  9.90 step/s, accuracy=0.97, loss=0.11, step=84000]
Valid: 100% 6944/6944 [00:21<00:00, 327.15 uttr/s, accuracy=0.81, loss=0.84]
Train:   0% 1/2000 [00:00<03:28,  9.61 step/s, accuracy=0.81, loss=0.55, step=84001]

average loss = 0.3661       average accuracy = 0.9035)


Train:  50% 1001/2000 [01:36<02:05,  7.97 step/s, accuracy=0.97, loss=0.09, step=85001]

average loss = 0.3494       average accuracy = 0.9084)


Train: 100% 2000/2000 [03:22<00:00,  9.89 step/s, accuracy=0.88, loss=0.52, step=86000]
Valid: 100% 6944/6944 [00:20<00:00, 338.21 uttr/s, accuracy=0.82, loss=0.81]
Train:   0% 2/2000 [00:00<05:07,  6.50 step/s, accuracy=0.91, loss=0.46, step=86002]

average loss = 0.3502       average accuracy = 0.9076)


Train:  50% 1002/2000 [01:34<02:00,  8.25 step/s, accuracy=0.97, loss=0.21, step=87001]

average loss = 0.3370       average accuracy = 0.9102)


Train: 100% 2000/2000 [03:15<00:00, 10.22 step/s, accuracy=0.97, loss=0.18, step=88000]
Valid: 100% 6944/6944 [00:21<00:00, 325.88 uttr/s, accuracy=0.82, loss=0.80]
Train:   0% 2/2000 [00:00<04:35,  7.25 step/s, accuracy=0.91, loss=0.27, step=88002]

average loss = 0.3344       average accuracy = 0.9107)


Train:  50% 1001/2000 [01:38<02:05,  7.93 step/s, accuracy=0.91, loss=0.58, step=89001]

average loss = 0.3348       average accuracy = 0.9124)


Train: 100% 2000/2000 [03:22<00:00,  9.87 step/s, accuracy=0.94, loss=0.33, step=9e+4]
Valid: 100% 6944/6944 [00:20<00:00, 335.74 uttr/s, accuracy=0.82, loss=0.80]
Train:   0% 1/2000 [00:00<05:39,  5.89 step/s, accuracy=0.91, loss=0.30, step=9e+4]

Step 90000, best model saved. (accuracy=0.8230)
average loss = 0.3301       average accuracy = 0.9126)


Train:  50% 1001/2000 [01:39<02:00,  8.28 step/s, accuracy=0.84, loss=0.42, step=91001]

average loss = 0.3230       average accuracy = 0.9162)


Train: 100% 2000/2000 [03:19<00:00, 10.04 step/s, accuracy=0.78, loss=0.52, step=92000]
Valid: 100% 6944/6944 [00:20<00:00, 335.73 uttr/s, accuracy=0.82, loss=0.81]
Train:   0% 2/2000 [00:00<04:42,  7.08 step/s, accuracy=0.91, loss=0.31, step=92002]

average loss = 0.3263       average accuracy = 0.9149)


Train:  50% 1001/2000 [01:41<02:02,  8.15 step/s, accuracy=0.84, loss=0.52, step=93001]

average loss = 0.3167       average accuracy = 0.9163)


Train: 100% 2000/2000 [03:20<00:00,  9.96 step/s, accuracy=0.91, loss=0.28, step=94000]
Valid: 100% 6944/6944 [00:21<00:00, 319.34 uttr/s, accuracy=0.83, loss=0.79]
Train:   0% 1/2000 [00:00<03:35,  9.28 step/s, accuracy=0.94, loss=0.16, step=94001]

average loss = 0.3118       average accuracy = 0.9184)


Train:  50% 1001/2000 [01:44<01:58,  8.43 step/s, accuracy=0.81, loss=0.93, step=95001]

average loss = 0.3193       average accuracy = 0.9164)


Train: 100% 2000/2000 [03:23<00:00,  9.82 step/s, accuracy=0.97, loss=0.23, step=96000]
Valid: 100% 6944/6944 [00:20<00:00, 331.19 uttr/s, accuracy=0.82, loss=0.77]
Train:   0% 2/2000 [00:00<05:14,  6.35 step/s, accuracy=0.97, loss=0.17, step=96002]

average loss = 0.3152       average accuracy = 0.9174)


Train:  50% 1002/2000 [01:41<02:10,  7.65 step/s, accuracy=0.88, loss=0.47, step=97002]

average loss = 0.3162       average accuracy = 0.9189)


Train: 100% 2000/2000 [03:17<00:00, 10.15 step/s, accuracy=1.00, loss=0.12, step=98000]
Valid: 100% 6944/6944 [00:20<00:00, 332.16 uttr/s, accuracy=0.82, loss=0.79]
Train:   0% 1/2000 [00:00<03:32,  9.39 step/s, accuracy=0.84, loss=0.38, step=98001]

average loss = 0.3088       average accuracy = 0.9190)


Train:  50% 1002/2000 [01:42<02:13,  7.46 step/s, accuracy=1.00, loss=0.10, step=99001]

average loss = 0.3088       average accuracy = 0.9196)


Train: 100% 2000/2000 [03:16<00:00, 10.17 step/s, accuracy=0.91, loss=0.41, step=1e+5]
Valid: 100% 6944/6944 [00:20<00:00, 336.33 uttr/s, accuracy=0.82, loss=0.78]
Train:   0% 0/2000 [00:00<?, ? step/s]


Step 100000, best model saved. (accuracy=0.8298)


######original : val acc = 68%
######d_model 80->128 , multihead 2->1
###### train acc = 80% , val acc = 73%

######TransformerEncoder layer : 1 -> 2 :

######TransformerEncoder layer : 1 -> 4 :
######train acc = 92% , val acc = 83%

In [20]:
import os
import json
import torch
from pathlib import Path
from torch.utils.data import Dataset


class InferenceDataset(Dataset):
  def __init__(self, data_dir):
    # using / to merge path
    testdata_path = Path(data_dir) / "testdata.json"
    metadata = json.load(testdata_path.open())
    self.data_dir = data_dir
    #
    self.data = metadata["utterances"]

  def __len__(self):
    return len(self.data)

  def __getitem__(self, index):
    utterance = self.data[index]
    feat_path = utterance["feature_path"]
    mel = torch.load(os.path.join(self.data_dir, feat_path))
    # compare : when we load training set, we use append to concate datas
    return feat_path, mel


def inference_collate_batch(batch):
  """Collate a batch of data."""
  feat_paths, mels = zip(*batch)

  return feat_paths, torch.stack(mels)

In [21]:
import json
import csv
from pathlib import Path
from tqdm.notebook import tqdm

import torch
from torch.utils.data import DataLoader

def parse_args():
  """arguments"""
  config = {
    "data_dir": "./Dataset",
    "model_path": "./model.ckpt",
    "output_path": "./output.csv",
  }

  return config


def main(
  data_dir,
  model_path,
  output_path,
):
  """Main function."""
  device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
  print(f"[Info]: Use {device} now!")

  mapping_path = Path(data_dir) / "mapping.json"
  mapping = json.load(mapping_path.open())

  dataset = InferenceDataset(data_dir)
  dataloader = DataLoader(
    dataset,
    batch_size=1,
    shuffle=False,
    drop_last=False,
    num_workers=8,
    collate_fn=inference_collate_batch,
  )
  print(f"[Info]: Finish loading data!",flush = True)

  speaker_num = len(mapping["id2speaker"])
  model = Classifier(n_spks=speaker_num).to(device)
  model.load_state_dict(torch.load(model_path))
  model.eval()
  print(f"[Info]: Finish creating model!",flush = True)

  results = [["Id", "Category"]]
  for feat_paths, mels in tqdm(dataloader):
    with torch.no_grad():
      mels = mels.to(device)
      outs = model(mels)
      preds = outs.argmax(1).cpu().numpy()
      # zip from two list to a sequence of tuples
      # preds is a numpy array whose elements are int numbers
      for feat_path, pred in zip(feat_paths, preds):
        # mapping["id2speaker"] = {"0": "id10001", ......'599':  'id11251'}])
        results.append([feat_path, mapping["id2speaker"][str(pred)]])

  with open(output_path, 'w', newline='') as csvfile:
    writer = csv.writer(csvfile)
    writer.writerows(results)


if __name__ == "__main__":
  main(**parse_args())


[Info]: Use cuda now!
[Info]: Finish loading data!
[Info]: Finish creating model!


/usr/local/lib/python3.10/dist-packages/torch/utils/data/dataloader.py:557: UserWarning: This DataLoader will create 8 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(


  0%|          | 0/6000 [00:00<?, ?it/s]

In [ ]:
class MyLinearLayer(nn.Module):
    """ Custom Linear layer but mimics a standard linear layer """
    def __init__(self, size_in, size_out):
        super().__init__()
        self.size_in, self.size_out = size_in, size_out
        weights = torch.Tensor(size_out, size_in)
        self.weights = nn.Parameter(weights)  # nn.Parameter is a Tensor that's a module parameter.
        bias = torch.Tensor(size_out)
        self.bias = nn.Parameter(bias)

        # initialize weights and biases
        nn.init.kaiming_uniform_(self.weights, a=math.sqrt(5)) # weight init
        fan_in, _ = nn.init._calculate_fan_in_and_fan_out(self.weights)
        bound = 1 / math.sqrt(fan_in)
        nn.init.uniform_(self.bias, -bound, bound)  # bias init

    def forward(self, x):
        w_times_x= torch.mm(x, self.weights.t())
        return torch.add(w_times_x, self.bias)  # w times x + b
class MyLinear(nn.Module):
    def __init__(self, in_units, units):
        super().__init__()
        self.weight = nn.Parameter(torch.randn(in_units, units))
        self.bias = nn.Parameter(torch.randn(units,))

    def forward(self, X):
        linear = torch.matmul(X, self.weight.data) + self.bias.data
        return F.relu(linear)
class NoisyLinear(T.nn.Module):
  def __init__(self, n_in, n_out):
    super().__init__()
    self.n_in, self.n_out = n_in, n_out

    self.weights = T.nn.Parameter(T.zeros((n_out, n_in),
      dtype=T.float32))
    self.bias = T.nn.Parameter(T.tensor(n_out,
      dtype=T.float32))
    self.lo = 0.90; self.hi = 0.98  # noise

    lim = 0.01  # initialize weights and bias
    T.nn.init.uniform_(self.weights, -lim, +lim)
    T.nn.init.uniform_(self.bias, -lim, +lim)

  def forward(self, x):
    wx= T.mm(x, self.weights.t())
    rnd = (self.hi - self.lo) * T.rand(1) + self.lo
    return rnd * T.add(wx, self.bias)  # wts * x + bias
class PrintLayer(nn.Module):
    def __init__(self):
        super(PrintLayer, self).__init__()

    def forward(self, x):
        # Do your print / debug stuff here
        print(x)
        return x

In [23]:
class PermuteLayer(nn.Module):
  def __init__(self, input, output):
    super().__init__()
  def forward(self , x):
    return(x.permute(1,0))

class SelfAttentionPollingLayer(nn.Module):
  def __init__(self,d_ap):
    super().__init__()
    self.attention_layer = \
    nn.Sequential(
        PermuteLayer(),
        nn.Linear(input, d_ap),
        nn.ReLU(),
        nn.Linear(d_ap, 1 ),
        nn.Softmax()
    )
  def forward(self ,x ):
    return torch.mm(x, self.attention_layer(x))